### PR TITLE
refactor(web): partial encapsulation of the desktop title bar and resize bar 🍕

### PR DESCRIPTION
**Pull request by jahorton**
This is a first pass for relocating code specific to the desktop form factor of the OSK.

Note that while it is currently based upon the branch of #5387, it can easily be rebased against `master`.  The code changes should be completely orthogonal for now.

In reference to the [current design doc for the OSK-Core initiative](https://docs.google.com/document/d/1uV2IgYv7RQUT318nkJ6Cjwgx8lhGfmtWPqPAD7myfIU/edit?usp=sharing), this PR begins the process of splitting off the `TargetedFloatLayout` aspects of the current `OSKManager` class into a separate class.

That said, there's enough within said target class (from the spec) to actually merit modularizing it.  As there's a lot of ground to cover, a relatively easy starting spot is the title and resize bars - they're both relatively isolated and reasonable to encapsulate.  There is still some code spaghetti (because event handlers) needed to keep things working in the current state of KMW, though.

For this pass, we're mostly concerned with encapsulating the DOM structure used to construct both bars.  The larger event handlers will remain in their current locations for the moment, as there's already quite a lot of code relocation taking place.  

There are quite a number of them to relocate eventually.  This is not an exhaustive list, but...
- All the `_VResize` functions are specific to the desktop layout
- All the `_VMoveMouse` functions are also specific to the desktop layout

Both of these have dedicated class fields that will need moving, which in turn means a lot more code-spaghetti to track.  So yeah, leaving that for another pass.

## Other Notes

Note that there will be changes "before" vs "after" for `sil_euro_latin` - the 'pin', configuration, close, and help buttons will all display now, unlike before.  I consider the "close" button to be the biggest win; I can now hide the "help" via the OSK.

**Before**

Screen Shot 2021-06-29 at 2 49 22 PM

**After**

Screen Shot 2021-06-29 at 2 46 54 PM

Also, the "KeymanWeb" part of the title doesn't seem to be the most consistent; I could spend extra effort there if we want it preserved perfectly.  (I opted for preserving the common case, which doesn't include it for keyboards with a standard OSK.)

No resize bar, as the OSK needs to ensure the help text is properly displayed.

------

I also noted the issue behind #5392 while testing out this refactor.  That one _is_ against `master`, as it was quite simple to isolate.

------

## User Testing
@keymanapp/testers 

_**Note**_: give preference to #5409, which builds directly on this one and affects the same areas.  You should be able to handle both sets of tests on #5409's branch without issue.

Testing should be performed against desktop browsers; the standard 'unminified' testing page should be fine.

Suggested keyboards:
- a basic US English keyboard
    - if using the unminified testing page, `English - English`
    - if not, I'd advise `basic_kbdus`. ("US Basic")
- `sil_euro_latin` (which has no OSK for the desktop form-factor, as seen above)
    - Note that the error behind #5392 will currently occur on this PR's branch if using Developer mode while testing.

### CODE REVIEW

File: web/source/osk/oskManager.ts [hunk 1]
```diff
@@ -357,123 +353,14 @@ namespace com.keyman.osk {
...
       } else {
         this.vkbd.adjustHeights(this);
       }
     }
 
-    /**
-     * Create a control bar with title and buttons for the desktop OSK
-     */
-    controlBar(): HTMLDivElement {
-      // TODO: merge with _TitleBarInterior?
```
Comment (jahorton) [inline web/source/osk/oskManager.ts:371]: Challenge accepted.  There's only one version (a blend of the two) in the final result.

File: web/source/osk/oskManager.ts [hunk 2]
```diff
@@ -346,9 +338,11 @@ namespace com.keyman.osk {
       // Set box class - OS and keyboard added for Build 360
       this._Box.className=util.device.formFactor+' '+ util.device.OS.toLowerCase() + ' kmw-osk-frame';
 
+      const layout = this.desktopLayout = new layouts.TargetedFloatLayout();
```
Comment (mcdurdin) [inline web/source/osk/oskManager.ts:341]: Should we be creating this if we aren't a 'desktop' formFactor? (see line 344)

File: web/source/osk/oskManager.ts [hunk 3]
```diff
@@ -1170,17 +993,17 @@ namespace com.keyman.osk {
...
         }
 
       }
       // Fix or release user dragging
       if('nomove' in p) {
         this.noDrag=p['nomove'];
-        if(this.pinImg) {
-          this.pinImg.style.display=(p['nomove'] || !this.userPositioned) ? 'none' : 'block';
+        if(this.desktopLayout) {
+          this.desktopLayout.titleBar.showPin(!(p['nomove'] || !this.userPositioned));
```
Comment (mcdurdin) [inline web/source/osk/oskManager.ts:1006]: This test is overly negative 😁 

```suggestion
          this.desktopLayout.titleBar.showPin(!p['nomove'] && this.userPositioned);
```

**Review by mcdurdin — APPROVED** [review]
This seems pretty straightforward compared to the last one 😄 

LGTM -- please allow user testing to go through before merge

**Comment by jahorton** [comment]
@keymanapp/testers

**Comment by MakaraSok** [comment]
Tested using http://localhost/keymanweb/testing/unminified.html when standing at:

```
makara@AORAL MINGW64 /c/keyman/keyman/web (refactor/web/osk-desktop-header-footer)
$ git status
On branch refactor/web/osk-desktop-header-footer
Your branch is up to date with 'origin/refactor/web/osk-desktop-header-footer'.

nothing to commit, working tree clean
```

Results:
1. the OSK can be repositioned by dragging on the title bar - **pass**
2. the OSK can be resized by clicking and dragging on the bottom right corner - **pass**

**From the user point of view**, two things that one might expected: (1) can the orange bar at the bottom be clicked and dragged to reposition the keyboard? and (2) it would be ideal that the keyboard layout keeps its dimension when resized -- we may not want to see something like below. They look ugly and not quite usable.

image

image

_Note also that the help icon doesn't take the user to the help page of the keyboard, let alone the KMW configuration options._

**Comment by jahorton** [comment]
> Tested using http://localhost/keymanweb/testing/unminified.html when standing at:
> 
> ```
> makara@AORAL MINGW64 /c/keyman/keyman/web (refactor/web/osk-desktop-header-footer)
> $ git status
> On branch refactor/web/osk-desktop-header-footer
> Your branch is up to date with 'origin/refactor/web/osk-desktop-header-footer'.
> 
> nothing to commit, working tree clean
> ```
> 
> Results:
> 
>     1. the OSK can be repositioned by dragging on the title bar - **pass**
> 
>     2. the OSK can be resized by clicking and dragging on the bottom right corner - **pass**
> 
> 
> **From the user point of view**, two things that one might expected: (1) can the orange bar at the bottom be clicked and dragged to reposition the keyboard? and (2) it would be ideal that the keyboard layout keeps its dimension when resized -- we may not want to see something like below. They look ugly and not quite usable.
> 
> image

By "keeps its dimension", do you mean its [aspect ratio](https://en.wikipedia.org/wiki/Aspect_ratio_(image))?  I'll admit that the screenshot above does have a rather nasty aspect ratio.

Or did the keyboard spontaneously 'pop' to a different size than what you had set it to before releasing the resize handle, like was seen with #5296?  Here's a video (from that issue) highlighting a 'pop' scenario, though while resizing occurred:

https://user-images.githubusercontent.com/4498365/122194512-db27ab80-ced8-11eb-991b-e76078d75ff2.mp4

**Comment by mcdurdin** [comment]
> By "keeps its dimension", do you mean its [aspect ratio](https://en.wikipedia.org/wiki/Aspect_ratio_(image))? I'll admit that the screenshot above does have a rather nasty aspect ratio. But, if that's where the user left it, we assume that's what the user wants.

Correct -- people often want to be able to place the keyboard somewhere "out of the way" which means forcing it into non-photogenic aspect ratios. We won't market the screenshots of those (ha ha because we do marketing do we?) but we will continue to make it possible to do that.

**Comment by MakaraSok** [comment]
> By "keeps its dimension", do you mean its [aspect ratio](https://en.wikipedia.org/wiki/Aspect_ratio_(image))? I'll admit that the screenshot above does have a rather nasty aspect ratio. But, if that's where the user left it, we assume that's what the user wants.

Not specific to this screenshot, but it would be nice to be able to log the aspect ratio (perfectly sized by default) to avoid having something like the above. I like the way the OSK on Windows and macOS works when it comes to resizing.

**Comment by mcdurdin** [comment]
> Not specific to this screenshot, but it would be nice to be able to lock the aspect ratio (perfectly sized by default) to avoid having something like the above. I like the way the OSK on Windows and macOS works when it comes to resizing.

I wonder if we could do this with holding shift while resizing or something like that? I don't want to fully remove the ability to stretch the keyboard because it can be handy in some situations on a web page. But this is a future feature idea ⛽

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.85-alpha](https://keyman.com/downloads/releases/alpha/15.0.85)

**Pull request merged**